File: src/presentation/galaxyState.ts
import { generateGalaxy } from "../galaxy/generateGalaxy.ts";
import type { Galaxy } from "../galaxy/generateGalaxy.ts";
import type { Planet } from "../data/types/planet.ts";
import { content, saveSystem } from "./gameState.ts";
import { ensureBootstrapColonization, withPlanetOwnership } from "./planetOwnershipState.ts";

// Phase 2 integration (Agent 10): replaces the MVP's single hardcoded
// Delta Rigelus with a real generated galaxy. Persistence goes through
// SaveSystem exclusively, per the architectural mandate -- only the SEED
// is stored (small, and generateGalaxy() is deterministic given it), not
// the whole planet array.
const GALAXY_SEED_KEY = "profitable:galaxySeed";

// Alpha Section 3 (docs/profitable-alpha-scale-performance-plan.md):
// locked at 50, the real alpha galaxy size -- large enough that the map
// genuinely can't be taken in at a glance, round enough to reason about
// when tuning discovery pacing/travel distances/market pool distribution
// against it. Was a small fixed 5 through Phase 2-5 and the deferred
// gaps, deliberately kept minimal until scale/performance became this
// section's own explicit job to verify -- see that plan doc for the full
// rationale and the four verification tests this change is gated behind.
const PLANET_COUNT = 50;

interface LoadedGalaxy {
  galaxy: Galaxy;
  isNewGalaxy: boolean;
}

function loadOrCreateGalaxy(): LoadedGalaxy {
  const storedSeed = saveSystem.load(GALAXY_SEED_KEY) as string | null;
  const galaxy = generateGalaxy(PLANET_COUNT, content.resources, storedSeed ?? undefined);
  const isNewGalaxy = !storedSeed;
  if (isNewGalaxy) {
    saveSystem.save(GALAXY_SEED_KEY, galaxy.seed);
  }
  return { galaxy, isNewGalaxy };
}

const loaded: LoadedGalaxy = loadOrCreateGalaxy();
export const galaxy: Galaxy = loaded.galaxy;
const isNewGalaxy: boolean = loaded.isNewGalaxy;

// Agent 8 always sets discovered: false on every generated planet (its own
// contract says picking/revealing a starting planet is "Agent 10's
// integration concern, not this agent's"). This is that concern.
const rawStartingPlanet = galaxy.planets[0]!;

// Phase 5 integration (Agent 22): travel needs at least one real reachable
// destination to demonstrate against. Mirrors startingPlanet's own
// override immediately above, one planet further into the generated list.
const rawSecondaryDiscoveredPlanet: Planet | undefined = galaxy.planets[1];

// Colonist-Driven Production bootstrap exception (planet-ownership.md):
// applied at the same moment a genuinely NEW seed is generated -- the only
// place that can tell "brand new galaxy" apart from "existing save
// reloading." Floor-set (ensureBootstrapColonization()'s own guarantee),
// so this is safe to be a no-op on every subsequent reload of the same
// save.
if (isNewGalaxy) {
  const bootstrapIds = [rawStartingPlanet.id, rawSecondaryDiscoveredPlanet?.id].filter(
    (id): id is string => id !== undefined,
  );
  ensureBootstrapColonization(bootstrapIds);
}

export const startingPlanet: Planet = withPlanetOwnership({ ...rawStartingPlanet, discovered: true });

export const secondaryDiscoveredPlanet: Planet | undefined = rawSecondaryDiscoveredPlanet
  ? withPlanetOwnership({ ...rawSecondaryDiscoveredPlanet, discovered: true })
  : undefined;

// Bug fix (Galactic Map Agent 25/26 verification, Section 6/7 of
// profitable-map-gdd.md): arriving at a planet via a real Voyage never
// used to mark it discovered anywhere in the codebase -- only the two
// bootstrap overrides above were ever reachable, regardless of how much a
// player traveled. The map GDD's own premise ("a planet becomes
// discovered: true when physically visited") is the design this closes.
//
// `galaxy.planets` itself only round-trips its SEED through SaveSystem
// (see loadOrCreateGalaxy() above) -- generatePlanet() deterministically
// reproduces discovered: false for every planet on every reload, so an
// in-memory-only mutation of galaxy.planets would silently vanish on the
// next session. Discovery-by-travel therefore needs its own persisted
// side-table, the same "elsewhere" pattern tradingState.ts's
// listingQualities already uses for data a regenerated/reloaded structure
// can't itself carry.
const DISCOVERED_PLANET_IDS_SAVE_KEY = "profitable:discoveredPlanetIds";

let discoveredPlanetIds: string[] =
  (saveSystem.load(DISCOVERED_PLANET_IDS_SAVE_KEY) as string[] | null) ?? [];

const bootstrapDiscoveredPlanets: Planet[] = [startingPlanet, secondaryDiscoveredPlanet].filter(
  (planet): planet is Planet => planet !== undefined,
);
const bootstrapDiscoveredIds = new Set(bootstrapDiscoveredPlanets.map((planet) => planet.id));

// The single source of truth for "which planets can the player currently
// see/travel to" -- the two structural bootstrap planets plus any real
// generated planet a Voyage has actually delivered a ship to. Presentation
// code (TradeMapScene) should read this rather than re-deriving its own
// copy of the discovery rule.
//
// Necessary correction (found while implementing the Scanner/Probe
// presentation amendment): traveledTo entries came straight off
// galaxy.planets, which always carries discovered: false baked in
// (generatePlanet() is deterministic and never mutated in place --
// membership in discoveredPlanetIds is what actually signals discovery,
// not the object's own field). That went unnoticed until performScan()
// needed a real, trustworthy Planet.discovered to check -- every planet
// this function returns is now normalized to discovered: true, matching
// what its own name and contract already promised.
//
// Colonist-Driven Production: every returned planet is also merged with
// its planetOwnershipState entry -- the same "single source of truth,
// merge at read time" discipline extended to ownership, not just
// discovery.
export function getDiscoveredPlanets(): Planet[] {
  const traveledTo = galaxy.planets
    .filter((planet) => !bootstrapDiscoveredIds.has(planet.id) && discoveredPlanetIds.includes(planet.id))
    .map((planet) => withPlanetOwnership({ ...planet, discovered: true }));
  return [...bootstrapDiscoveredPlanets, ...traveledTo];
}

// Called from a successful resolveArrival() (physical visitation) and,
// since the Scanner/Probe amendment, from a successful performScan() too
// (TradeMapScene.onScan()) -- both are physical/deliberate discovery
// paths the map GDD's "no alternate discovery path" guard (see
// tests/integration/mapVerification.test.ts) still confirms are the only
// two: no code anywhere else writes a Planet's `discovered` field.
export function markPlanetDiscovered(planetId: string): void {
  if (bootstrapDiscoveredIds.has(planetId) || discoveredPlanetIds.includes(planetId)) return;
  discoveredPlanetIds = [...discoveredPlanetIds, planetId];
  saveSystem.save(DISCOVERED_PLANET_IDS_SAVE_KEY, discoveredPlanetIds);
}

File: unity/ProfitableUnity/Assets/Scripts/Content/GalaxyState.cs
#nullable enable
using Profitable.Core.Constants;
using Profitable.Core.Schema;
using Profitable.Core.Simulation;

namespace Profitable.Unity.Content
{
    // Migration Phase 2 Sub-Phase A (Galaxy, Planet, Mining) Presentation/
    // Integration -- docs/agents/agent-41-unity-galaxy-planet-presentation.md.
    // Parallels GameContent.cs's own lazy-static shape but stays a separate
    // class: GameContent is scoped to the raw content catalog (Agent 35's
    // contract), this owns the generated galaxy built on top of it.
    //
    // Session-only, in-memory, no persistence -- the same deliberate scope
    // limit Inventory.cs already draws (full save-system integration is a
    // later sub-phase's job, not this one's). A fresh galaxy generates once
    // per process and is reused for the rest of the session; nothing here
    // claims to survive a restart the way src/presentation/galaxyState.ts's
    // real seed-persisted version does.
    public static class GalaxyState
    {
        // Matches src/presentation/galaxyState.ts's real alpha-scale
        // PLANET_COUNT -- already parity-tested at this exact size
        // (agent-40-unity-galaxy-planet-parity-validation.md's galaxyCases),
        // not an arbitrary smaller demo number.
        private const int PlanetCount = 50;

        // Fixed rather than randomly generated per session: this class has
        // no persistence, so a random seed would make the galaxy (and any
        // test asserting on it) different every process run. Reproducible
        // until a later sub-phase ports real seed persistence through
        // ISaveSystem.
        private const string Seed = "unity-mvp-galaxy";

        private static Galaxy? _galaxy;

        public static Galaxy Galaxy => _galaxy ??= Generate();

        // galaxy.Planets[0], overridden exactly as galaxyState.ts overrides
        // its own rawStartingPlanet:
        // - Discovered forced true. PlanetGenerator.Generate() always
        //   emits false ("picking/revealing a starting planet is a later
        //   agent's concern" -- agent-39's own comment); this is that
        //   agent.
        // - ColonistCount floored to the production minimum. A direct,
        //   in-memory analog of galaxyState.ts's ensureBootstrapColonization()
        //   floor exception, scoped down to skip the full persisted
        //   planetOwnershipState side-table, which is Sub-Phase E's own job
        //   to port. Without this override the Colonist-Driven Production
        //   gate (ported early as PlanetOwnershipConstants
        //   .MinimumColonistsToProduce) would block the MVP gather loop
        //   before Sub-Phase E's real ownership system exists.
        public static Planet StartingPlanet => Galaxy.Planets[0];

        // Migration Phase 2 Sub-Phase D (Ships & Travel) addition --
        // agent-56-unity-ships-travel-presentation.md. Mirrors the real
        // src/presentation/galaxyState.ts's own secondaryDiscoveredPlanet
        // exactly: "travel needs at least one real reachable destination
        // to demonstrate against... one planet further into the generated
        // list," forced Discovered=true the same way the starting planet
        // already is. Unlike the starting planet, no colonist floor is
        // applied here -- this planet exists only as a travel
        // destination, not a second gatherable location (that would be
        // Sub-Phase E's own colonization scope to extend).
        public static Planet SecondaryDestinationPlanet => Galaxy.Planets[1];

        private static Galaxy Generate()
        {
            var galaxy = GalaxyGenerator.Generate(PlanetCount, GameContent.Loaded.Resources, Seed);
            var startingPlanet = galaxy.Planets[0];
            startingPlanet.Discovered = true;
            startingPlanet.ColonistCount = PlanetOwnershipConstants.MinimumColonistsToProduce;
            galaxy.Planets[1].Discovered = true;
            return galaxy;
        }

        // Mirrors GameContent.ResetForTests()'s own hook -- a test that
        // wants a fresh galaxy after asserting on the static cache can
        // force one.
        public static void ResetForTests() => _galaxy = null;
    }
}
